barretenberg/cpp/src/barretenberg/ecc/batched_affine_addition/batched_affine_addition.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include "barretenberg/ecc/curves/bn254/bn254.hpp"
+#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
+#include <cstddef>
+#include <cstdint>
+
+namespace bb {
+
+template <typename Curve> class BatchedAffineAddition {
+    using G1 = typename Curve::AffineElement;
+    using Fr = typename Curve::ScalarField;
+    using Fq = typename Curve::BaseField;
+
+    // Struct describing a set of points to be reduced to num-sequence-counts-many points via summation of each sequence
+    struct AdditionSequences {
+        std::vector<size_t> sequence_counts;
+        std::span<G1> points;
+        std::span<Fq> scratch_space;
+    };
+
+    // Collection of addition sequences to be handled by each thread
+    struct ThreadData {
+        std::vector<AdditionSequences> addition_sequences;
+        std::vector<std::vector<size_t>> sequence_tags; // allows for the recombining of sequences split across threads
+    };
+
+  public:
+    /**
+     * @brief Given a set of points and sequence counts, peform addition to reduce each sequence to a single point
+     * @details Reduce each sequence to a single point via repeated rounds of pairwise addition. (If the length
+     * of the sequence is odd in a given round, the unpaired point is simply carried over to the next round). The
+     * inverses needed in the addition formula are batch computed in a single go for all additions to be performed
+     * across all sequences in a given round.
+     * @note: Multithreading is achieved by evenly distributing the points across the optimal number of available
+     * threads. This can result in the bisecting of some sequences which is acounted for in the final result by further
+     * summing the reduced points that resulted from a sequence split across two or more threads. An example with two
+     * threads and three add sequences:
+     *
+     *                     |------------------------| Points
+     *                     |------------|-----------| Thread boundaries
```

Comment:
Should the second/interior `|` on the `Thread Boundaries` line align with the third `|` on the `New addition sequence boundaries` line?

Reply:
yes, thanks